docs/design/interpreter/precise-gc.md

```diff
@@ -0,0 +1,23 @@
+### Synchronisation with GC suspension
+
+Interpreter will run in GC unsafe mode. This means that suspension can happen only at certain points in the interpreter code. certain.
+* Explicit safepoints. We need to ensure that the interpreter can enter a gc safe state in a very short time. For this purpose, we will insert explicit safepoint instructions, most likely at method entry and as part of a backwards branch. The safepoint will poll whether there is a pending interruption request, in which case it will save any information that is needed by the stack walker so the frames can be processed. (for example the current `ip` will need to be flushed to the `InterpFrame` and the `InterpFrame` will need to be pushed on some tls stack to serve as the starting point of unwinding)
+* Locations where we end up calling into the runtime. For example, as part of executing an opcode, we might need to call into the runtime to ensure that the class is initialized and the cctor is being run. The same rules as for explicit safepoints should apply, by saving any information needed by the stack walker.
+
+Interpreter instruction typically read values from the interpreter stack, then proceed with the operation and in the end it might end some value back to the interpreter stack. For more complex opcodes that call into the runtime, a GC could happen in the middle of the opcode execution. All these locations where GC interruption can happen should be well defined, because we will have to enforce the rule that every object reference is refetched from the interpreter stack following a potential suspention point. Otherwise we could end up with native registers/stack containing an old ref for a moved objects. References that are being used by the opcode could also be stored in a separate memory location associated to the `GCFrame` of the native interpreter execution loop method.
```

Comment:
```suggestion
Interpreter instruction typically read values from the interpreter stack, then proceed with the operation and in the end it might end some value back to the interpreter stack. For more complex opcodes that call into the runtime, a GC could happen in the middle of the opcode execution. All these locations where GC interruption can happen should be well defined, because we will have to enforce the rule that every object reference is refetched from the interpreter stack following a potential suspension point. Otherwise we could end up with native registers/stack containing an old ref for a moved objects. References that are being used by the opcode could also be stored in a separate memory location associated to the `GCFrame` of the native interpreter execution loop method.
```